components/eam/src/chemistry/modal_aero/aero_model.F90

```diff
@@ -2083,76 +2099,75 @@ subroutine calc_schmidt_number( freepath, r_aer, temp,        & ! in
-          if(associated(fldcw)) then
-             do k=1,pver
-                vmr(:ncol,k,m) = mbar(:ncol,k) * fldcw(:ncol,k) / adv_mass(m)
-             enddo
-          else
-             vmr(:,:,m) = 0.0_r8
-          endif
+    integer  :: kk, mm
+    real(r8) :: fldcw(ncol,pver) ! mass mixing ratio [kg/kg]
+
+    do mm=1,gas_pcnst
+       if( adv_mass(mm) /= 0._r8 ) then
+          fldcw = fldcw_all(:,:,mm)
+          do kk=1,pver
+             vmr(:ncol,kk,mm) = mbar(:ncol,kk) * fldcw(:ncol,kk) / adv_mass(mm)
+          enddo
        endif
     enddo
   end subroutine qqcw2vmr
 
 
   !=============================================================================
   !=============================================================================
-  subroutine vmr2qqcw( lchnk, vmr, mbar, ncol, im, pbuf )
+  subroutine vmr2qqcw( vmr, mbar, ncol, fldcw_all )
     !-----------------------------------------------------------------
     !	... Xfrom from volume to mass mixing ratio
+    ! C++ porting: this subroutine is similar to the subroutine vmr2mmr 
+    ! in mozart/mo_mass_xforms.F90. Maybe can merge them
+    ! Note that the subroutine needs adv_mass from module chem_mod, 
+    ! but the values are assigned from module mo_sim_dat
     !-----------------------------------------------------------------
 
-    use m_spc_id
-    use modal_aero_data, only : qqcw_get_field
-
     implicit none
 
     !-----------------------------------------------------------------
     !	... Dummy args
     !-----------------------------------------------------------------
-    integer, intent(in)     :: lchnk, ncol, im
-    real(r8), intent(in)    :: mbar(ncol,pver)
-    real(r8), intent(in)    :: vmr(ncol,pver,gas_pcnst)
-    type(physics_buffer_desc), pointer :: pbuf(:)
+    integer, intent(in)     :: ncol  ! indexes
+    real(r8), intent(in)    :: mbar(ncol,pver) ! mean wet atmospheric mass [g/mol]
+    real(r8), intent(in)    :: vmr(ncol,pver,gas_pcnst) ! volume mixing ratios [mol/mol]
+    real(r8), intent(out)   :: fldcw_all(ncol,pver,gas_pcnst) ! mass mixing ratio [kg/kg]
 
     !-----------------------------------------------------------------
     !	... Local variables
     !-----------------------------------------------------------------
-    integer :: k, m
-    real(r8), pointer :: fldcw(:,:)
+    integer  :: kk, mm
     !-----------------------------------------------------------------
     !	... The non-group species
     !-----------------------------------------------------------------
-    do m = 1,gas_pcnst
-       fldcw => qqcw_get_field(pbuf, m+im,lchnk,errorhandle=.true.)
-       if( adv_mass(m) /= 0._r8 .and. associated(fldcw)) then
-          do k = 1,pver
-             fldcw(:ncol,k) = adv_mass(m) * vmr(:ncol,k,m) / mbar(:ncol,k)
+    fldcw_all(:,:,:) = 0.0_r8
+    do mm = 1,gas_pcnst
+       if( adv_mass(mm) /= 0._r8) then
+          do kk = 1,pver
+             fldcw_all(:ncol,kk,mm) = adv_mass(mm) * vmr(:ncol,kk,mm) / mbar(:ncol,kk)
```

Comment:
I think it will be easier to remove the `ncol` argument in this routine. Please do so if it is not too troublesome.